File: MNIST/kNN.cpp
#include <iostream>
#include <thread>
#include <fstream>
#include <string>
#include <vector>
#include <regex>
#include <algorithm>

#include <typeinfo>

using namespace std;

string test_filename  = "test.csv";
string train_filename = "train.csv";

int main() {
	ifstream test_file(test_filename);
	ifstream train_file(train_filename);

	string header, train_point, test_point;
	test_file >> header;
	train_file >> header;

	// test_file >> test_point;
	// cout << test_point;
	while(train_file >> train_point) {
		vector<int> tp_vector;
		regex commas(",");
		sregex_token_iterator iter(train_point.begin(), train_point.end(), commas, -1);
		sregex_token_iterator done;

		for(; iter != done; ++iter) {
			tp_vector.push_back(stoi(iter->str()));
		}
	}

	// vector<vector<int>> test_data;
	// vector<vector<int>> train_data;

	return 1;
}

File: MNIST/kNN.py
import csv
from collections import Counter
# import math for sqrt
import numpy as np
# import pyspark for distance calculations later

#WARNING: This script is VERY SLOW!

# Dimensions of a and b expected to be 1xN
# Euclidean Distance function: 
# d = sum (a[i] - b[i])^2
def dist(train_data, test_data):
	return sum((train_data - test_data) ** 2)

k = 3

testfile = open('test.csv', 'rb')
trainfile = open('train.csv', 'rb')

train_data = csv.reader(trainfile)
test_data = csv.reader(testfile)

train_data.next()
test_data.next()

train = []
test = []
for row in train_data:
	new_row = []
	for val in row:
		new_row.append(int(val))
	train.append(np.array(new_row))


sumbissionfile = open('submission.csv', 'w')
sumbissionfile.write("ImageId,Label\n")
test_number = 1
for row in test_data:
	tp = []
	for val in row:
		tp.append(int(val))
	test_point = np.array(tp)

	distances = []
	for train_point in train:
		label = train_point[0]
		d = dist(train_point[1:], test_point)
		distances.append( [d, train_point[0]] )
	sorted_distances = np.argsort(distances, axis=0)

	labels = []
	for i in xrange(k):
		label = distances[sorted_distances[i][0]][1]
		labels.append(label)

	sumbissionfile.write(str(test_number))
	sumbissionfile.write(",")
	sumbissionfile.write(str(Counter(labels).most_common(1)[0][0]))
	sumbissionfile.write("\n")
	test_number = test_number + 1
